### PR TITLE
Add validation of OpenAPI 2.0 spec

### PR DESCRIPTION
**Pull request by pderaaij**
In our use case we are looking for a solution where we can validate 2.0 and 3.0 specifications and our custom linting rules. I like the setup of this project, but it misses the 2.0 validation. As an exercise to see how hard it would be I started to experiment. It turned out to be rather easy and this PR adds the ability to validate against 2.0 specification as well. Letting 3.0 being the default validation.

I am aware that this project is targeted towards 3.0 specification. If this PR is missing out on the goals of the project I understand, but it feels to me that it could be a useful addition.

Opening this PR to share my work so far and start discussing if this is a wanted addition.

### CODE REVIEW

File: lib/validation-strategies/openapi-v2.js [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+}
+
+function validateSync(openapi, options, callback) {
+    setupOptions(options,openapi);
+
+    should(openapi).be.an.Object();
+    openapi.should.not.have.key('openapi');
+    openapi.should.have.key('swagger');
+    openapi.swagger.should.have.type('string');
+    should.ok(openapi.swagger.startsWith('2.'), 'Must be an OpenAPI 2.x document');
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: Should test for exactly `2.0` only.

File: lib/validation-strategies/openapi-v2.js [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+    openapi.should.have.key('paths');
+    should(openapi.paths).be.an.Object();
+
+    for (let k in openapi) {
+        if (!k.startsWith('x-')) {
+            should(['swagger', 'info', 'host', 'basePath', 'schemes', 'consumes','produces','paths', 'definitions','parameters','responses', 'securityDefinitions', 'security', 'tags', 'externalDocs'].indexOf(k)).be.greaterThan(-1,'OpenAPI object cannot have additionalProperty: '+k);
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: Are the `produces` and `consumes` arrays themselves validated?

File: lib/validation-strategies/openapi-v2.js [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+        schema.type.should.equalOneOf('integer','number','string','boolean','object','array'); // not null
+        if (schema.type === 'array') {
+            schema.should.have.property('items');
+        }
+    }
+    if (schema.allOf) {
+        schema.allOf.should.be.an.Array();
+        schema.allOf.should.not.be.empty();
+    }
+    if (schema.anyOf) {
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: `anyOf`, `oneOf` and `not` were not present in OpenAPI 2.0

File: lib/validation-strategies/openapi-v2.js [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+        this.errors = params.errors;
+    }
+}
+
+function setupAjv(ajv) {
+    const ajvFormats = require('ajv/lib/compile/formats.js');
+    ajv.addFormat('uriref', ajvFormats.full['uri-reference']);
+    ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
+    ajv._refs['http://json-schema.org/schema'] = 'http://json-schema.org/draft-04/schema'; // optional, using unversioned URI is out of spec
+    const metaSchema = require('ajv/lib/refs/json-schema-v5.json');
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: OpenAPI 2.0 used a version of JSON Schema based on Draft 4, so some of this Draft 5 setup is not required.

Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: OpenAPI 2.0 used a version of JSON Schema based on Draft 4, so some of this Draft 5 setup is not required.

File: lib/validation-strategies/openapi-v2.js [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+        var ref = response.$ref;
+        should(response.$ref).be.type('string');
+        lintObject('reference',response,options);
+        response = common.resolveInternal(openapi, ref);
+        should(response).not.be.exactly(false, 'Cannot resolve reference: ' + ref);
+    }
+    response.should.have.property('description');
+    should(response.description).have.type('string', 'response description should be of type string');
+
+    response.should.not.have.property('schema', 'response schema must go inside `content` and be listed under a mime type');
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: Doesn't sound right for OpenAPI 2.0

File: lib/validation-strategies/openapi-v2.js [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+        openapi.info.license.name.should.have.type('string');
+        if (typeof openapi.info.license.url !== 'undefined') {
+            should.doesNotThrow(function () { validateUrl(openapi.info.license.url, 'license.url', options) },'Invalid license.url');
+        }
+        lintObject('license', openapi.info.license, options);
+        options.context.pop();
+    }
+
+    if (typeof openapi.info.termsOfService !== 'undefined') {
+        should.doesNotThrow(function () { validateUrl(openapi.info.termsOfService, 'termsOfService', options) },'Invalid termsOfService.url');
```
Comment (MikeRalphson) [inline lib/validation-strategies/openapi-v2.js]: `termsOfService` did not need to be a URL in OpenAPI 2.0

File: lib/validation-strategies/openapi-v2.js (hunk 2 above)
Comment (pderaaij) [reply lib/validation-strategies/openapi-v2.js]: I've added validation of the produces and consumes in my last commit.

**Comment by pderaaij** [comment]
Definitely understand the concerns and love to have a look to make this more futureproof.

I'll take a look at the resources shared and see what I can do. Will give an update when I have something to show.

**Comment by pderaaij** [comment]
@philsturgeon Just to figure out what's possible I've created a proof of concept. shared in this gist: https://gist.github.com/pderaaij/a3ca921118bbe9298c8d359c0730dd63

Of course it is far from perfect, but it shows it can validate schemes using the OpenAPI json scheme as a reference. Is this in line what you are thinking?

I was able to successfully validate tests/validator/pass/OAI/petstore.yaml and successfully invalidate a few of the fail files.

**Comment by MikeRalphson** [comment]
@pderaaij see also https://github.com/wework/speccy/blob/master/lib/walkSchema.js

One of the intentions of the current hand-written 3.0.x validator was to provide better and more informative error messages than are achievable with a validator based on a JSON schema instance. The validator also checks for constraints which cannot currently be expressed by JSON Schema draft 4/5.

The validator currently uses `ajv` as a fallback to validate the instance against one of the draft OpenAPI 3.0 schemas, and I know @philsturgeon has been toying with the idea of possibly reversing that so schema validation is done first - possibly in conjunction with use of [better-ajv-errors](https://github.com/torifat/better-ajv-errors). If this is done, obviously using the OpenAPI 2.0 schema for validation would be trivial.

The rich validator would then need only to check things that `ajv` cannot. A substantial amount of code could be removed.

I'm still not sure what is the best approach to maintain the benefits of a rich validator but apply it to multiple versions of the OpenAPI specification.

**Comment by philsturgeon** [comment]
Hey, I left this open for ages because I was very conflicted. At some point we'll need to support v3.1 and v4.0 but I would rather tackle that as we go. As @MikeRalphson pointed out, right now there is a _lot_ of code which handles "rich validation" and as we move more towards schema validation first a lot of that code will go away. 

As and when 3.1/4.0 features pop up, we can look into featuring multiple versions. But, as we've moved this logic to oas-kit now anyway, the whole thing is a bit moot. Sorry! :)

**Pull request closed without merging**